### PR TITLE
Lmtp multircptstatus

### PR DESCRIPTION
**Pull request by mschneider82**
added support for LMTP multi rcpt status 
Also added TCP Support, works also with starttls

### CODE REVIEW

**Comment by mschneider82** [comment]
Link to #23

**Comment by mschneider82** [comment]
I found a issue when then same rcpt is used multiple times in a Mail. I will add code next Werk to tempfail If rcpt is already in []reciepients

File: conn.go [hunk 1]
```diff
@@ -476,17 +495,61 @@ func (c *Conn) handleData(arg string) {
...
+	s.rcptStatus[rcpt].ch <- status
+}
+
+func (s *lmtpStatusCollector) StartDelivery(ctx context.Context, rcpt string) {
+	rcpt = strings.ToLower(rcpt)
+	s.rcptStatus[rcpt] = &rcptStatus{
+		ch:  make(chan *SMTPError),
+		ctx: ctx,
+	}
+	s.wg.Add(1)
```
Comment (emersion) [inline conn.go]: I don't get why this is necessary. When `Session.Data` returns the delivery of all receipients is complete. I don't understand why it's necessary to break this assumption.

File: README.md [hunk 2]
```diff
@@ -93,7 +93,7 @@ func (s *Session) Rcpt(to string) error {
 	return nil
 }
 
-func (s *Session) Data(r io.Reader) error {
+func (s *Session) Data(r io.Reader, sc smtp.LMTPStatusCollector) error {
```
Comment (emersion) [inline README.md]: Since this is only for LMTP, I wonder if there's a way to make it optional…

File: conn.go (hunk 1 above)
Comment (mschneider82) [reply conn.go]: When you are writing a lmtp server which for example stores every rcpt mails to a s3 bucket, you can do this parallel in go routines in background and once it's finished you can reply the code to the client

File: README.md (hunk 2 above)
Comment (mschneider82) [reply README.md]: Only if we would split out lmtp completely. But its not hard for SMTPD to fill that interface, just update the data function definition

File: conn.go (hunk 1 above)
Comment (emersion) [reply conn.go]: You can just do the wait group stuff in the `Data` command.

Comment (mschneider82) [reply conn.go]: The wait group is not needed anymore I will remove it

Comment (mschneider82) [reply conn.go]: You need a way to tell for every rcpt the status at the end of Data. I think the easiest way to solve this it the extended interface. Even with ctx and cancel/timeout support. As lib user you don't care about the logic just use the sc interface, very easy

Comment (emersion) [reply conn.go]: I don't get why `StartDelivery` is necessary at all. You just need to tell go-smtp when a recipient has succeeded/failed.

Comment (mschneider82) [reply conn.go]: It's needed when you want to reply the SMTPError code immediately to the lmtpclient.

When you wait until all rcpts are finished the client may be timed out and will retry the mail. Lmtp is to prevent this, once it's delivered to a backend the client needs the status immediately

Comment (mschneider82) [reply conn.go]: Please look into server_test.go how-to use it. Assume there is a http upload in the go routine or maildir delivery

Comment (emersion) [reply conn.go]: >It's needed when you want to reply the SMTPError code immediately to the lmtpclient.

Then just return the error from the function.

I'd rather not change the fact that the delivery is finished exactly when the function returns. It makes everything a lot more complicated.

**Comment by emersion** [comment]
LMTP replies should be in the same order as the `RCPT` commands. We should enforce this.

I'm starting to wonder whether it's worth it to go through all this trouble just to send replies as soon as possible. Maybe just a return value containing the status for each recipient is enough.

**Comment by mschneider82** [comment]
The replies are correctly ordered with my code

File: conn.go (hunk 1 above)
Comment (mschneider82) [reply conn.go]: Lmtp is different here the delivery for each rcpt is a own task. If you have many rcpts and a slow delivery to the backend its important to tell tell client about the rcpt status asap. In my case I have such a problem that's why I switched to lmtp instead of smtp

**Comment by emersion** [comment]
>The replies are correctly ordered since we range over recipients and ask the channels for the SMTPError in the correct order

This isn't enforced. The backend can send statuses in incorrect order, and go-smtp will mix everything up.

File: conn.go (hunk 1 above)
Comment (emersion) [reply conn.go]: Why is it important to tell clients ASAP?

I understand that it's important that the client knows which deliveries have succeeded or failed, but I'm not sure LMTP has been designed for notifying the client as soon as possible. The RFC requires recipient statuses to be sent in order, which somewhat makes that difficult.

Let's say you have 3 recipients, and the first one is very slow. You can't send the status for the completed 2 last recipients before the first one has completed.

**Comment by mschneider82** [comment]
the ordering is enforced, in line 498:

https://github.com/mschneider82/go-smtp/blob/lmtp-multircptstatus/conn.go#L498

we loop over the reciepents in the correct order, once there is no status the channel read blocks until the status is available.

**Comment by emersion** [comment]
>the ordering is enforced

Ah, indeed, missed that there was one channel per recipient.

File: conn.go (hunk 1 above)
Comment (mschneider82) [reply conn.go]: > Let's say you have 3 recipients, and the first one is very slow. You can't send the status for the completed 2 last recipients before the first one has completed.

Yes thats correct, the final status response is always as fast as the first rcpt delivery, this depends on the ordering in RFC, but we collect the status asap and deliver them to the client

Comment (emersion) [reply conn.go]: Then I'd prefer to just return a value with the status for each recipient.

**Comment by mschneider82** [comment]
Yes every rcpt has a chan for his status, or if there is a timeout or deadline the ctx will abort it and tempfails the recipient.

I am sorry that this breaks the existing api, the only way to make this would be to splitt out lmtp to a different interface and not using the smtp one. But this duplicates much code

File: conn.go (hunk 1 above)
Comment (mschneider82) [reply conn.go]: In this case the lib user has todo the ordering correctly and all the channels and context handling, then it would be better to split out lmtp to a own interface.

The other issue is that if you have 50 rcpts and a big mail the delivery can take every long, the lmtp client may times out before we finished all rcpts. In this case the lmtpclient will retry every 50 rcpt (even if there where already 48rcpts successful delivered to the backend).

File: README.md (hunk 2 above)
Comment (foxcpp) [reply README.md]: If we are really serious about backward-compatibility - we can add an interface like that:
```go
type LMTPSession interface {
  LMTPData(r io.Reader, sc smtp.LMTPStatusCollector) error
}
```
and assert for it.

But I agree with mschneider82, it is pretty easy to fix existing code and we are at 0.x.

File: backend.go [hunk 3]
```diff
@@ -33,5 +34,10 @@ type Session interface {
 	// Add recipient for currently processed message.
 	Rcpt(to string) error
 	// Set currently processed message contents and send it.
-	Data(r io.Reader) error
+	Data(r io.Reader, s LMTPStatusCollector) error
+}
+
+type LMTPStatusCollector interface {
```
Comment (foxcpp) [inline backend.go]: Question. Why can't we have it as a structure? I can't think of a use case where that interface would be useful.

File: conn.go [hunk 4]
```diff
@@ -476,17 +495,58 @@ func (c *Conn) handleData(arg string) {
...
+			rcptStatus := statuscollector.rcptStatus[rcpt]
+			select {
+			case <-rcptStatus.ctx.Done():
+				c.Server().ErrorLog.Printf("Context Error: %s - tempfailing", rcptStatus.ctx.Err())
+				tempfail := SMTPError{
+					Code:         420,
+					EnhancedCode: EnhancedCode{4, 4, 7},
+					Message:      "Error: timeout reached",
+				}
+				status = &tempfail
```
[the remark this answers is not in the document]
Comment (mschneider82) [reply conn.go]: sure i can refactor, i just wanted to print out that its a tempfail for easier readable

File: conn.go [hunk 5]
```diff
@@ -325,6 +326,23 @@ func (c *Conn) handleRcpt(arg string) {
...
+		sliceContains := func(s []string, str string) bool {
+			for _, a := range s {
+				if a == str {
+					return true
+				}
+			}
+			return false
+		}
+
+		if sliceContains(c.recipients, strings.ToLower(recipient)) {
```
[the remark this answers is not in the document]
Comment (mschneider82) [reply conn.go]: a second solution would be to create a seperate map for it, i can do this

File: conn.go [hunk 6]
```diff
@@ -476,17 +495,58 @@ func (c *Conn) handleData(arg string) {
...
+		rcptStatus: make(map[string]*rcptStatus),
+	}
+}
+
+func (s *lmtpStatusCollector) SetStatus(rcpt string, status *SMTPError) {
+	rcpt = strings.ToLower(rcpt)
+	s.rcptStatus[rcpt].ch <- status
+}
+
+func (s *lmtpStatusCollector) StartDelivery(ctx context.Context, rcpt string) {
```
Comment (emersion) [inline conn.go]: Still not sure why calling this is necessary.

File: README.md (hunk 2 above)
Comment (emersion) [inline README.md]: Could rename to a neutral name like `smtp.DataContext` for future extensibility.

File: conn.go (hunk 6 above)
Comment (mschneider82) [reply conn.go]: its required to have a start point to create the channel and also assign the ctx, each rcpt can have a own ctx child with seperate timeout/deadline

File: README.md (hunk 2 above)
Comment (mschneider82) [reply README.md]: ok fine!

File: conn.go (hunk 6 above)
Comment (emersion) [reply conn.go]: Can we leave the context for later? There are too many new things in this PR.

>create the channel

We can do that when creating the `lmtpStatusCollector`

File: conn.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (mschneider82) [reply conn.go]: DONE with map (see new pull req)

File: conn.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (mschneider82) [reply conn.go]: DONE (see new pull req)

File: README.md (hunk 2 above)
Comment (mschneider82) [reply README.md]: DONE with map (see new pull req)

**Comment by mschneider82** [comment]
Ok everything refactored:

introduced a map for rcpt duplicate checks (better performace)
refactored do DataContext
handled the other nitpicking

File: server_test.go [hunk 7]
```diff
@@ -86,6 +88,23 @@ func (s *session) Data(r io.Reader) error {
 			s.backend.messages = append(s.backend.messages, s.msg)
 		}
 	}
+
+	ctx, _ := context.WithTimeout(context.Background(), 3*time.Second)
+	for _, rcpt := range s.msg.To {
+		rcpt := rcpt
+		sc.StartDelivery(ctx, rcpt)
+		go func() {
```
[the remark this answers is not in the document]
Comment (mschneider82) [reply server_test.go]: once go routine per RcptData delivery seems fine to me, its the best performace, but this highly depends on the Data() implementation, the code also supports sequential processing.

File: conn.go (hunk 6 above)
Comment (mschneider82) [reply conn.go]: when creating lmtpStatusCollector/ (refactored to dataContext), we dont have a ctx per rcpt. Without a ctx we cannot implement a timeout, but this is highly needed

Comment (emersion) [reply conn.go]: 1. Isn't one context for all recipients enough?
2. Why do we need to add this to go-smtp? Can't backends implement a timeout internally?
3. As said above, I'm not a fan of "big pull requests" adding a lot of new features.

Comment (mschneider82) [reply conn.go]: currently we dont have a context support in the library, There should be a way to get the ctx from the library user ( i think StartDelivery is a elegant way, you can use the same ctx as in a http library ), its not forced to use a ctx for each rcpt, but its possible

it isnt that big pull request, I rebased it into one commit, we just add the feature "multi rcpt status", of couse with ctx/channel support, if a user doesnt want to parallel delivery to his backend he just have to not use go() routines in his code. we collected the status just with channels thats the go way.

**Comment by mschneider82** [comment]
I have added README example code for lmtp

File: server_test.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (foxcpp) [reply server_test.go]: Alright.

File: server.go [hunk 8]
```diff
@@ -147,15 +146,15 @@ func (s *Server) handleConn(c *Conn) error {
 // ListenAndServe listens on the network address s.Addr and then calls Serve
 // to handle requests on incoming connections.
 //
-// If s.Addr is blank and LMTP is disabled, ":smtp" is used.
```
Comment (emersion) [inline server.go]: This change doesn't belong to this PR.

Comment (mschneider82) [reply server.go]: If we dont support TCP for lmtp, then the README.md example with telnet is wrong.

Comment (emersion) [reply server.go]: Why would it be wrong? The current README example creates a TCP server.

Comment (mschneider82) [reply server.go]: With my patch i have enabled TCP mode for LMTP, this was not allowed before and was forced to unixsocket. 
I have created an example for a LMTP TCP server with telnet as client in README.md, this example would be wrong if we dont accept TCP Mode for lmtp here.

**Comment by mschneider82** [comment]
Lets discuss how we can go ahead here, I made much effort to get this ready and tested. 

Currently the LMTP support this in library is missing the "lmtp" feature to accept/not accept single rcpts, because of this feature lmtp was designed.

My patch is of couse breaking the existing API, but there is no way to not break it, because the current api is too simple for this kind of usage. 

Returning a slice of []error is breaking the RFC because it should be "near real-time". (also may breaks the api). A other solution sould be to call Data() for each rcpt but this is too strange.

I refactored the new interfaace to a general DataContext, this may can be used in future also for SMTP (e.g. context cancelation, getting all []recipients or something else).

**Comment by emersion** [comment]
First of all, I appreciate your work on this PR, thanks for this.

I'm fine with breaking the API. However here are the things I dislike with this PR:

* Does too much. TCP for LMTP should be discussed separately. Context support should be added separately.
* `StartDelivery` is unnecessary
* I'd like to keep the README small, I don't want to clutter it with another server example.

**Comment by mschneider82** [comment]
Ok, let's talk about 'StartDelivery is unnecessary': we could create the rcpt channels silent in the backend but then we have no possibility to add the ctx to it. The ctx is important to not block the channels and have a second way to get a status. 

I actually had other ways implemented and tested, also with performance tests

**Comment by mschneider82** [comment]
I found another use case for the `DataContext`, i have implemented [XFORWARD](http://www.postfix.org/XFORWARD_README.html) support in my [feature branch](https://github.com/mschneider82/go-smtp/commits/feature-xforward)

Through the DataContext you can get the XForward informations from the mail.

**Comment by foxcpp** [comment]
>I found another use case for the DataContext, i have implemented XFORWARD support in my feature branch

Side note: Since XFORWARD info doesn't change during transaction, it can be passed in ConnectionState structure to AnonymousLogin/Login.

Btw, what's the status of this PR?

I agree with emersion concerns about LMTP/TCP and README.
Can you revert both changes?

Also I believe that context support can be dropped altogether. I don't see any direct use for it and would like to keep library interfaces free from unnecessary details. Also, I think context-based timeouts and cancellation can be implemented in the backend without any help from the library.

**Comment by mschneider82** [comment]
> The XFORWARD attributes are reset after the MAIL FROM transaction completes (after RSET or DOT), so there is no risk of information leakage.

**Comment by mschneider82** [comment]
I suggest that i close this pull request, since my solution is not compatible with your vision of the library. I have to live with my fork. I'd like to keep the implementation of a server simple and not the library.

With your design every implementation of a lmtp server needs to handle with channels and status, timeouts (also the order of the status replies), this is too advanced in my opinion for an implementation.

Please use this as a base of better lmtp support. you can also check my master branch for other ideas (like the embbeded smtp.DefaultSession implementation)

Thanks for the library and your ideas.

**Comment by foxcpp** [comment]
Closing due to previous comment by @mschneider82. I will submit "good" bits as my own PR.
